umbrella-feeds/src/lib.rs

```diff
@@ -0,0 +1,730 @@
+        CustomContractError::NotEnoughSignatures
+    );
+
+    let mut prev_signer = AccountAddress([0u8; 32]);
+
+    let message_hashes = contract_view_message_hash(ctx, host, crypto_primitives)?;
+
+    let mut validators: Vec<AccountAddress> = vec![];
+
+    // To save gas we check only required number of signatures.
+    // The case, where you can have part of signatures invalid but still enough valid in total is not supported.
+    for i in 0..host.state().required_signatures {
+        let signer = param.signers_and_signatures[i as usize].0;
+        let signature = &param.signers_and_signatures[i as usize].1;
+
+        //Check signature.
+        let valid_signature =
+            host.check_account_signature(signer, signature, &message_hashes[i as usize])?;
+        ensure!(valid_signature, CustomContractError::WrongSignature);
+
+        ensure!(
+            prev_signer < signer,
+            CustomContractError::SignaturesOutOfOrder
+        );
+
+        validators.push(signer);
+
+        prev_signer = signer;
+    }
+
+    let are_valid_signers = host.invoke_contract_read_only::<Vec<AccountAddress>>(
+        &host.state().staking_bank,
+        &validators,
+        EntrypointName::new_unchecked("verifyValidators"),
+        Amount::zero(),
+    )?;
+
+    let are_valid_signers: bool = are_valid_signers
+        .ok_or(CustomContractError::InvokeContractError)?
+        .get()?;
+
+    ensure_eq!(are_valid_signers, true, CustomContractError::InvalidSigner);
+
+    Ok(())
+}
+
+#[receive(
+    contract = "umbrella_feeds",
+    name = "update",
+    parameter = "UpdateParams",
+    error = "CustomContractError",
+    crypto_primitives,
+    mutable
+)]
+fn update<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    crypto_primitives: &impl HasCryptoPrimitives,
+) -> Result<(), CustomContractError> {
+    let param: UpdateParams = ctx.parameter_cursor().get()?;
+
+    let message = param.message;
+
+    // Check that the signature was intended for this contract.
+    ensure_eq!(
+        message.contract_address,
+        ctx.self_address(),
+        CustomContractError::WrongContract
+    );
+
+    // Check signature is not expired.
+    ensure!(
+        message.timestamp > ctx.metadata().slot_time(),
+        CustomContractError::Expired
+    );
+
+    // Check signature has correct chain_id.
+    ensure_eq!(
+        message.chain_id,
+        CHAIN_ID,
+        CustomContractError::ChainIdMismatch
+    );
+
+    verify_signatures(ctx, host, crypto_primitives)?;
+
+    for element in message.price_feed {
+        let price_key: HashSha2256 = element.0;
+        let price_data: PriceData = element.1;
+
+        // We do not allow for older prices.
+        // This prevents replay attacks by preventing reusing of signatures at the same time.
+        let old_price_data = host.state().prices.get(&price_key).map(|s| *s);
```

Comment:
Use [`StateMap::entry`](https://docs.rs/concordium-std/latest/concordium_std/struct.StateMap.html#method.entry) when you need to first read and then write, since this will keep a reference to the entry, removing the need to pay for a key lookup again when inserting below.